### PR TITLE
Report performance of MIOpen with tuned and untuned MLIR kernels

### PR DESCRIPTION
**Pull request by yiqian1**
- Jenkinsfile:
  Add stages for tuning MLIR kernels and running MIOpen with the user perfdb. Tuning and testing are on different nodes. 

- MIOpenDriver.py
   Add a command option "-bmiopen_use_mlir" for running MIOpenDriver with MLIR kernels

- createPerformanceReports.py
   Create html for performance results

- perfRegressionReport.py
- reportUtils.py
   Apply style to tuned/untuned results

- SqliteDb.cpp
   Has nothing to do with this PR, but I accidentally found the syntax errors in this file, so I fixed them.

The CI job is here:
http://ml-ci.amd.com:21096/job/MLIR/job/MLIR-tuned-kernel-perf/48/

### CODE REVIEW

File: mlir/utils/jenkins/Jenkinsfile [hunk 1]
```diff
@@ -256,47 +256,149 @@ pipeline {
...
                         }
-                        stage("Build MLIR") {
+                        stage("Tune MLIR Kernels") {
                             steps {
-                                buildProject('mlir-miopen-driver mlir-rocm-runner ci-performance-scripts', '')
+                                buildProject('ci-performance-scripts', '')
+                                dir('build') {
+                                    sh 'python3 ./bin/MIOpenDriver.py -tuning'
+                                }
+                                sh 'ls -alr MIOpen/build; ls -alr /root;'
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile]: What's all this for?

File: mlir/utils/jenkins/Jenkinsfile [hunk 2]
```diff
@@ -256,47 +256,149 @@ pipeline {
...
+                                   -DCMAKE_INSTALL_PREFIX=${WORKSPACE}/MIOpen/build/MIOpenInstallDir
+                                   ''')
+                                   sh 'cd build; make -j $(nproc)'
+                                   sh 'cd build; make install'
+                               }
+                                stash name:"MIOpen-kernels",\
+                                   includes: "MIOpen/build/**"
+                           }
+                        } 
+                        stage("Build libMLIRMIOpen and MLIR") {
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile:273]: I noticed we have a bunch of rebuilds of the project - can we look for ways to reuse these builds instead?

File: mlir/utils/jenkins/Jenkinsfile (hunk 1 above)
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile]: That is, please remove the debug `ls` calls before merge.

File: mlir/utils/jenkins/Jenkinsfile [hunk 3]
```diff
@@ -256,47 +256,149 @@ pipeline {
...
+                    environment {
+                        // Make libMIOpen.so accessible to the test driver
+                        LD_LIBRARY_PATH="${WORKSPACE}/MIOpen/build/lib:$LD_LIBRARY_PATH:"
+                   }
+                   stages {
+                        stage("Copy MIOpen with MLIR kernels") {
+                            steps {
+                                sh 'rm -rf MIOpen'
+                                unstash name: "MIOpen-MLIR-kernels"
+                                sh 'ls -al MIOpen/build'
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile]: Debug ls needs removed

File: mlir/utils/jenkins/Jenkinsfile [hunk 4]
```diff
@@ -256,47 +256,149 @@ pipeline {
...
+                        stage("Test Tuned vs Untuned") {
+                            steps {
+                                buildProject('ci-performance-scripts', '')
+                                sh 'ldconfig'
+                                dir('build') {
+                                   sh 'python3 ./bin/MIOpenDriver.py -bmiopen_use_mlir'
+                                   sh 'mv miopen_perf.csv miopen_tuned_perf.csv'
+                                   sh 'rm -rf MIOpen/build/MIOpenUserDB'
+                                   sh 'python3 ./bin/MIOpenDriver.py -bmiopen_use_mlir'
+                                   sh 'mv miopen_perf.csv miopen_untuned_perf.csv'
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile]: Instead of these `mv` calls, can we get an option to the performance tester to specify the output file?

File: mlir/utils/jenkins/Jenkinsfile [hunk 5]
```diff
@@ -306,7 +408,59 @@ pipeline {
...
+                    steps {
+                        buildProject('ci-performance-scripts', '')
+
+                        copyArtifacts filter: 'build/*.csv,build,build/perf-run-date',\
+                                               optional: true,\
+                                               flatten: true,\
+                                               projectName: "/${JOB_NAME}",\
+                                               selector: lastSuccessful(),\
+                                               target: 'build/oldData'
+                        sh 'ls -alrt build'
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile]: Same nit wrt `ls`

File: mlir/utils/jenkins/Jenkinsfile [hunk 6]
```diff
@@ -306,7 +408,59 @@ pipeline {
...
+                    }
+                }
 
+                stage("Create performance reports") {
+                    steps {
+                        dir('build/reports') {
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile]: There any reason we're not unstashing right into `build/`?

File: mlir/utils/jenkins/MIOpenDriver.py [hunk 7]
```diff
@@ -263,14 +263,15 @@ def runConfigWithMLIR(config):
...
         outs, errs = p2.communicate()
 
-def runConfigWithMIOpenDriver(commandLine):
+def runConfigWithMIOpenDriver(commandLine, envs):
     # remove the result file generated by rocprof in previous benchmarking
     os.system("rm "+BENCHMARKING_RESULT_FILE_NAME)
     MIOpenDriverCommand = MIOPEN_DRIVER + ' ' + ' '.join(commandLine) + ' -V 0'
     print("Running MIOpen Benchmark: ", commandLine)
     profilerCommand = ROCPROF + ' --stats ' + MIOpenDriverCommand
+    print(envs)
```
Comment (krzysz00) [inline mlir/utils/jenkins/MIOpenDriver.py]: If you want to leave this info in, please add some sort of "with environment" annotation so we know what it's doing.

File: mlir/utils/jenkins/MIOpenDriver.py [hunk 8]
```diff
@@ -263,14 +263,15 @@ def runConfigWithMLIR(config):
...
     # invoke rocprof + MIOpenDriver.
-    p1 = subprocess.Popen(profilerCommand.split(), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    p1 = subprocess.Popen(profilerCommand.split(), stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=envs)
```
Comment (krzysz00) [inline mlir/utils/jenkins/MIOpenDriver.py]: Sanity check: does that add to or overwrite the environment?

File: mlir/utils/jenkins/MIOpenDriver.py [hunk 9]
```diff
@@ -302,45 +303,75 @@ def benchmarkMIOpen(commandLine, xdlops):
...
-
-    with open("MLIR_vs_MIOpen.html", 'w') as htmlOutput:
-        reportUtils.htmlReport(df, means, "MLIR vs. MIOpen performance", "MLIR/MIOpen", htmlOutput)
-
-def generatePerformanceResults(configs, xdlops):
-    mlir_df = pd.DataFrame(benchmarkMLIR(testVector.split(sep=' '), xdlops) for testVector in configs)
-    miopen_df = pd.DataFrame(benchmarkMIOpen(testVector.split(sep=' '), xdlops) for testVector in configs)
-
-    printPerformance(mlir_df, miopen_df)
+    with pd.option_context('precision', reportUtils.ROUND_DIGITS):
```
Comment (krzysz00) [inline mlir/utils/jenkins/MIOpenDriver.py]: Is this also a leftover debug print?

File: mlir/utils/jenkins/MIOpenDriver.py [hunk 10]
```diff
@@ -302,45 +303,75 @@ def benchmarkMIOpen(commandLine, xdlops):
...
+def tuneMLIRKernels(configs, xdlops):
+    solver_names = {testVector : getSolverName(testVector, xdlops) for testVector in configs}
+
+    envs = dict(MIOPEN_FIND_ENFORCE='4', MIOPEN_DRIVER_USE_GPU_REFERENCE='1')
+    for testVector in configs:
+        envs['MIOPEN_DEBUG_FIND_ONLY_SOLVER']=solver_names[testVector]
+        commandLine = testVector.split(sep=' ')
+        config = ConvConfiguration.fromCommandLine(commandLine, xdlops)
+        if config.inputLayout == 'nchw':
+            MIOpenDriverCommand = MIOPEN_DRIVER + ' ' + ' '.join(commandLine) + ' -V 0'
```
Comment (krzysz00) [inline mlir/utils/jenkins/MIOpenDriver.py]: Instead of string append and then a split(), use a list of arguments. That is, `[MIOPEN_DRIVER, *commandLine, '-V', '0']` or something along those lines.

File: mlir/utils/jenkins/MIOpenDriver.py [hunk 11]
```diff
@@ -302,45 +303,75 @@ def benchmarkMIOpen(commandLine, xdlops):
...
+            print(MIOpenDriverCommand)
+            p1 = subprocess.Popen(MIOpenDriverCommand.split(), stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=envs)
```
Comment (krzysz00) [inline mlir/utils/jenkins/MIOpenDriver.py]: This join/split setup makes us vulnernerable to shell quoting shenanangins.

File: mlir/utils/jenkins/MIOpenDriver.py [hunk 12]
```diff
@@ -302,45 +303,75 @@ def benchmarkMIOpen(commandLine, xdlops):
...
+            # get output.
+            try:
+               outs, errs = p1.communicate(timeout=180)
```
Comment (krzysz00) [inline mlir/utils/jenkins/MIOpenDriver.py:358]: Since we're getting the output, can we slide a check in here to make sure we've output what's expected?

File: mlir/utils/jenkins/createPerformanceReports.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        df['Tuned/Untuned'] = df['MIOpen TFlops (Tuned MLIR Kernels)']/df['MIOpen TFlops (Untuned MLIR Kernels)']
+        df.to_csv(reportUtils.PERF_REPORT_FILE, index=False)
+        COLUMNS_TO_AVERAGE = ['MLIR TFlops', 'MIOpen TFlops', 'MLIR/MIOpen',
+                              'MIOpen TFlops (Tuned MLIR Kernels)', 'MIOpen TFlops (Untuned MLIR Kernels)', 'Tuned/Untuned']
+
+    plotMean = df[COLUMNS_TO_AVERAGE].agg(reportUtils.geoMean)
+    plotMean.name = "Geo. mean"
+    plotMean = pd.DataFrame(plotMean).T
+
+    plotMean[['MLIR TFlops', 'MIOpen TFlops']]\
```
Comment (krzysz00) [inline mlir/utils/jenkins/createPerformanceReports.py:47]: Do we want more columns here?

File: mlir/utils/jenkins/createPerformanceReports.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        .to_csv(reportUtils.PERF_PLOT_REPORT_FILE, index=False)
+
+    means = df.groupby(["Direction", "DataType"])[COLUMNS_TO_AVERAGE]\
+        .agg(reportUtils.geoMean)
+    means.loc[("All", "All"),:] = df[COLUMNS_TO_AVERAGE].agg(reportUtils.geoMean)
+    means.to_csv(reportUtils.PERF_STATS_REPORT_FILE)
+
+    with open("MLIR_vs_MIOpen.html", 'w') as htmlOutput:
+        reportUtils.htmlReport(df, means, "MLIR vs. MIOpen performance", "MLIR/MIOpen", htmlOutput)
```
Comment (krzysz00) [inline mlir/utils/jenkins/createPerformanceReports.py]: Missing list brackets given the changes to htmlReport?

**Review by krzysz00 — COMMENTED** [review]
I think there's a bunch of debug stuff left in here, mainly, and it might be a good idea to go looking for refactoring opportunities in the Jenkinsfile.

File: mlir/utils/jenkins/Jenkinsfile (hunk 2 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile:273]: I think so. Actually I'm working on it. There will be another PR.

File: mlir/utils/jenkins/Jenkinsfile (hunk 4 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile]: Done. Use the file names defined in reportUtils to remove coupling.

File: mlir/utils/jenkins/MIOpenDriver.py (hunk 10 above)
Comment (yiqian1) [reply mlir/utils/jenkins/MIOpenDriver.py]: Good point. Thanks.

File: mlir/utils/jenkins/MIOpenDriver.py (hunk 12 above)
Comment (yiqian1) [reply mlir/utils/jenkins/MIOpenDriver.py:358]: I used to have a check here, but it seems MIOpenDirver prints some log info to stderr, and the errs are quite confusing. So I decided not to have a check in here.

File: mlir/utils/jenkins/createPerformanceReports.py (hunk 13 above)
Comment (yiqian1) [reply mlir/utils/jenkins/createPerformanceReports.py:47]: Not now. We're more concerned about Tuned/Untuned.

File: mlir/utils/jenkins/createPerformanceReports.py (hunk 14 above)
Comment (yiqian1) [reply mlir/utils/jenkins/createPerformanceReports.py]: Yeah. I had updated it. Don't know how the change got lost. Fixed.

File: mlir/utils/jenkins/Jenkinsfile (hunk 2 above)
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile:273]: Cool!

File: mlir/utils/jenkins/MIOpenDriver.py (hunk 12 above)
Comment (krzysz00) [reply mlir/utils/jenkins/MIOpenDriver.py:358]: The validation tests have the check to make sure MIOpen ran correctly - maybe we can adapt that?

File: mlir/utils/jenkins/Jenkinsfile [hunk 15]
```diff
@@ -433,21 +427,18 @@ pipeline {
...
                     }
                 }
 
                 stage("Create performance reports") {
                     steps {
-                        dir('build/reports') {
+                        dir('build') {
                             unstash name: "Perf-report"
-                            sh 'ls'
                             sh 'mv build/mlir_vs_miopen_perf.csv ${WORKSPACE}/build'
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile:437]: Isn't this `mv` now a noop?

**Review by krzysz00 — COMMENTED** [review]
I think I just have a few more minor nitpicks - overall, thanks for putting in the work!

File: mlir/utils/jenkins/MIOpenDriver.py (hunk 12 above)
Comment (yiqian1) [reply mlir/utils/jenkins/MIOpenDriver.py:358]: Performance tests purposely turn off validation ('-V 0') to make more reasonable time results. I think we can assume MIOpen runs correctly here, as we also run validation tests separately.

File: mlir/utils/jenkins/Jenkinsfile (hunk 15 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile:437]: No. The unstashed files are at build/build/.

File: mlir/utils/jenkins/MIOpenDriver.py (hunk 12 above)
Comment (krzysz00) [reply mlir/utils/jenkins/MIOpenDriver.py:358]: My main concern here is making sure we don't accidentally compare against a no-op "run", but that can be a seperate PR.

File: mlir/utils/jenkins/Jenkinsfile [hunk 16]
```diff
@@ -306,7 +402,56 @@ pipeline {
...
+                                               projectName: "/${JOB_NAME}",\
+                                               selector: lastSuccessful(),\
+                                               target: 'build/oldData'
+                    }
+                }
+
+                stage("Create performance reports") {
+                    steps {
+                        dir('build') {
+                            unstash name: "Perf-report"
```
Comment (krzysz00) [inline mlir/utils/jenkins/Jenkinsfile:436]: Maybe move the `unstash` outside of the `dir('build')` block?

**Review by krzysz00 — APPROVED** [review]
Minor nitpick about `unstash` aside, looks good to me

I'm assuming you've run these tests by hand to make sure the report generation is successful?

File: mlir/utils/jenkins/MIOpenDriver.py (hunk 12 above)
Comment (yiqian1) [reply mlir/utils/jenkins/MIOpenDriver.py:358]: > Minor nitpick about `unstash` aside, looks good to me
> 
> I'm assuming you've run these tests by hand to make sure the report generation is successful?

Yes. The tests just finished.
http://ml-ci.amd.com:21096/job/MLIR/job/MLIR-tuned-kernel-perf/

File: mlir/lib/Dialect/MIOpen/Tuning/SqliteDb.cpp [hunk 17]
```diff
@@ -33,7 +33,7 @@ class SQLite::impl {
     } else {
       LLVM_DEBUG(dbgs() << "FATAL ERROR! Does not support user db\n");
     }
-    ptrDb = Sqlite3Ptr{ptr_tmp};
+    Sqlite3Ptr ptrDb = Sqlite3Ptr{ptr_tmp};
```
Comment (jerryyin) [inline mlir/lib/Dialect/MIOpen/Tuning/SqliteDb.cpp]: Please revert this change, you make a member variable to a local variable with same name.

File: mlir/utils/jenkins/Jenkinsfile [hunk 18]
```diff
@@ -256,47 +256,143 @@ pipeline {
...
+                                                     '''-DMIOPEN_BACKEND=HIP
+                                   -DCMAKE_PREFIX_PATH="${WORKSPACE}/MIOpen/MIOpenDeps"
+                                   -DCMAKE_INSTALL_PREFIX=${WORKSPACE}/MIOpen/build/MIOpenInstallDir
+                                   ''')
+                                   sh 'cd build; make -j $(nproc)'
+                                   sh 'cd build; make install'
+                               }
+                                stash name:"MIOpen-kernels",\
+                                   includes: "MIOpen/build/**"
+                           }
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile:271]: Please correct me if I'm wrong but I believe you need to clean up the cmake build cache (after stash) otherwise it will get used by next stage.

File: mlir/utils/jenkins/Jenkinsfile [hunk 19]
```diff
@@ -256,47 +256,143 @@ pipeline {
...
-                        stage("Build MLIR") {
+                        stage("Tune MLIR Kernels") {
                             steps {
-                                buildProject('mlir-miopen-driver mlir-rocm-runner ci-performance-scripts', '')
+                                buildProject('ci-performance-scripts', '')
+                                dir('build') {
+                                    sh 'python3 ./bin/MIOpenDriver.py -tuning'
+                                }
+                                stash name:"MIOpen-MLIR-kernels",\
+                                    includes: "MIOpen/build/**"
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile:300]: ~~Don't you need to stash the perfdb, not the build?~~ Ok, looks like you put the perfdb together with the build result.

~~Also, it looks like we are doing tuning in two independent stages:~~
 - ~~Tune MLIR Kernels~~
 - ~~MIOpen resnet50 config test~~

On a second look this also seems fine: The tuning stage do not check correctness whereas the later one does.

File: mlir/utils/jenkins/Jenkinsfile [hunk 20]
```diff
@@ -256,47 +256,143 @@ pipeline {
...
+                                unstash name: "MIOpen-MLIR-kernels"
+                            }
+                        }
+                        stage("Test Tuned vs Untuned") {
+                            steps {
+                                buildProject('ci-performance-scripts', '')
+                                sh 'ldconfig'
+                                dir('build') {
+                                   sh 'python3 ./bin/MIOpenDriver.py -bmiopen_use_tuned_mlir'
+                                   sh 'rm -rf MIOpen/build/MIOpenUserDB'
```
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile:389]: Looks good!

File: mlir/lib/Dialect/MIOpen/Tuning/SqliteDb.cpp (hunk 17 above)
Comment (yiqian1) [reply mlir/lib/Dialect/MIOpen/Tuning/SqliteDb.cpp]: I don't get it. The problem was that ptrDb was declared without a type. I just added the type.

File: mlir/utils/jenkins/Jenkinsfile (hunk 18 above)
Comment (yiqian1) [reply mlir/utils/jenkins/Jenkinsfile:271]: You're right. The build directory is deleted in getAndBuildMIOpen(), so we don't have to worry about it here.

File: mlir/lib/Dialect/MIOpen/Tuning/SqliteDb.cpp (hunk 17 above)
Comment (jerryyin) [reply mlir/lib/Dialect/MIOpen/Tuning/SqliteDb.cpp]: This is an assignment not a declaration. `ptrDb` is a member variable and in this line it got to assign to a value.

File: mlir/utils/jenkins/Jenkinsfile (hunk 18 above)
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile:271]: Ok gotcha

File: mlir/lib/Dialect/MIOpen/Tuning/SqliteDb.cpp (hunk 17 above)
Comment (yiqian1) [reply mlir/lib/Dialect/MIOpen/Tuning/SqliteDb.cpp]: Got it. The "ptrDb is not declared" error was actually caused by the missing parenthesis. I've reverted the change.

**Review by jerryyin — APPROVED** [review]
